ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/JsonQueryStreamResponseWriter.java

```diff
@@ -54,16 +72,34 @@ public class JsonQueryStreamResponseWriter implements QueryStreamResponseWriter
       = LoggerFactory.getLogger(JsonQueryStreamResponseWriter.class);
 
   private final HttpServerResponse response;
+  private final String mediaType;
+  private ConnectSchema connectSchema;
+  private KsqlConnectSerializer<Struct> serializer;
 
   @SuppressFBWarnings(value = "EI_EXPOSE_REP2")
-  public JsonQueryStreamResponseWriter(final HttpServerResponse response) {
+  public JsonQueryStreamResponseWriter(final HttpServerResponse response, final String mediaType) {
     this.response = Objects.requireNonNull(response);
+    this.mediaType = Objects.requireNonNull(mediaType);
   }
 
   @Override
   public QueryStreamResponseWriter writeMetadata(final QueryResponseMetadata metaData) {
     final Buffer buff = Buffer.buffer().appendByte((byte) '[');
-    buff.appendBuffer(ServerUtils.serializeObject(metaData));
+    if (mediaType.equals(KsqlMediaType.KSQL_V1_PROTOBUF.mediaType())) {
+      final LogicalSchema schema = metaData.schema;
+      final String queryId = metaData.queryId;
+
+      connectSchema = ConnectSchemas.columnsToConnectSchema(schema.columns());
+      serializer = new ProtobufNoSRSerdeFactory(ImmutableMap.of())
+              .createSerializer(connectSchema, Struct.class, false);
+
+      final StreamedRow header = StreamedRow.headerProtobuf(
```

Comment:
just checked here, we have `schema` anyway so it should be pretty cheap to just pipe it in right here